src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,458 @@
+                    Ok((src, dst, payload)) =>
+                        self.ingress(iface, now, payload, &src, &dst),
+                    Err(Error::Exhausted) =>
+                        (),
+                    Err(e) =>
+                        return Err(e),
+                }
+        }
+
+        // Send requests
+        if raw_socket.can_send() && now >= self.next_egress {
+            self.egress(iface, &mut *raw_socket, &checksum_caps, now)?;
+        }
+
+        Ok(())
+    }
+
+    fn ingress<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, now: Instant, data: &[u8], src: &IpEndpoint, dst: &IpEndpoint) {
+        if src.port != UDP_SERVER_PORT ||
+           dst.port != UDP_CLIENT_PORT { return }
+        let src_ip = match src.addr {
+            IpAddress::Ipv4(src_addr) => src_addr,
+            _ => return,
+        };
+
+        let dhcp_packet = match DhcpPacket::new_checked(data) {
+            Ok(dhcp_packet) => dhcp_packet,
+            Err(_) => return,
+        };
+        let dhcp_repr = match DhcpRepr::parse(&dhcp_packet) {
+            Ok(dhcp_repr) => dhcp_repr,
+            Err(_) => return,
+        };
+        let mac = iface.ethernet_addr();
+        if dhcp_repr.client_hardware_address != mac { return }
+        if dhcp_repr.transaction_id != self.transaction_id { return }
+        let server_identifier = match dhcp_repr.server_identifier {
+            Some(server_identifier) => server_identifier,
+            None => return,
+        };
+        net_debug!("DHCP recv {:?} from {} ({})", dhcp_repr.message_type, src, server_identifier);
+
+        if (dhcp_repr.message_type == DhcpMessageType::Offer ||
+            dhcp_repr.message_type == DhcpMessageType::Ack) &&
+           dhcp_repr.your_ip != Ipv4Address::UNSPECIFIED {
+               let prefix_len = dhcp_repr.subnet_mask
+                   .and_then(|mask| IpAddress::Ipv4(mask).to_prefix_len());
+               // Replace the first IP address
+               iface.update_ip_addrs(|addrs| {
+                   for cidr in addrs.iter_mut() {
+                       match (cidr.address(), prefix_len) {
+                           (IpAddress::Ipv4(_), Some(prefix_len)) => {
+                               let ipv4_cidr = Ipv4Cidr::new(dhcp_repr.your_ip, prefix_len);
+                               *cidr = IpCidr::Ipv4(ipv4_cidr);
+                               break
+                           }
+                           _ => ()
+                       }
+                   }
+               });
+               // Set gateway
+               match dhcp_repr.router {
+                   Some(router) if iface.in_same_network(&router.into()) => {
+                       let _ = iface.routes_mut()
+                           .add_default_ipv4_route(router.into());
+                   }
+                   _ => ()
+               }
+               // DNS servers
+               match dhcp_repr.dns_servers {
```

Comment:
I think returning a structure with the DHCP data, and maybe a method to autoconfigure the interface in a "simple" way would be ideal. Then you wouldn't have to store DNS servers here either.